src/Init/Data/BitVec/Lemmas.lean

```diff
@@ -1549,4 +1549,45 @@ theorem zeroExtend_truncate_succ_eq_zeroExtend_truncate_or_twoPow_of_getLsb_true
     simp [hx]
   · by_cases hik' : k < i + 1 <;> simp [hik, hik'] <;> omega
 
+@[simp]
+theorem replicate_zero_eq {x : BitVec w} : x.replicate 0 = 0#0 := by
+  simp [replicate]
+
+@[simp]
+theorem replicate_succ_eq {x : BitVec w} :
+    x.replicate (n + 1) =
+    (x ++ replicate n x).cast (by rw [Nat.mul_succ]; omega) := by
+  simp [replicate]
+
+/--
+If a number `w * n ≤ i < w * (n + 1)`, then `i - w * n` equals `i % w`.
+Intuitively, this is true by subtracting `w * n` from the inequality, giving
+`0 ≤ i - w * n < w`, which uniquely identifies `i % w`.
+-/
+private theorem Nat.sub_mul_eq_mod_of_lt_of_le (hlo : w * n ≤ i) (hhi : i < w * (n + 1)) :
+    i - w * n = i % w := by
+  rw [Nat.mul_succ] at hhi
+  rw [Nat.sub_eq_of_eq_add]
+  suffices i / w = n by rw [← this, Nat.add_comm, Nat.div_add_mod]
+  apply Nat.div_eq_of_lt_le
+    (by rw [Nat.mul_comm]; omega)
+    (by rw [Nat.mul_comm, Nat.mul_succ]; omega)
```

Comment:
This theorem should be factored into two statements:
```
example (m k : Nat) : m % k = m - k * (m / k) := by sorry
-- Surely in the library? The Int version is called `Int.emod_def`. `Nat.mod_add_div` is close.
```
```
example (hlo : w * n ≤ i) (hhi : i < w * (n + 1)) : n = i / w := ...
```
Both of these should then live in `Nat/Div.lean`.

Reply:
Resolved.

Reply:
`example (hlo : w * n ≤ i) (hhi : i < w * (n + 1)) : n = i / w := ...` already exists as `div_eq_of_lt_le`, so I use it directly.